File: crates/instrument/src/classes/async_switch.rs
use super::dialect_io;
use crate::classes::switch::parse_closed;
use instrument_core::error::Result;
use instrument_core::kind::InstrumentKind;
use instrument_core::scpi_commands;
use instrument_core::AsyncInstrumentSession;

/// Async switch / matrix session view (IVI-inspired / SCPI :ROUTe).
pub struct AsyncSwitch {
    session: AsyncInstrumentSession,
}

impl AsyncSwitch {
    pub fn new(session: AsyncInstrumentSession) -> Self {
        Self { session }
    }

    pub fn session(&self) -> &AsyncInstrumentSession {
        &self.session
    }

    pub fn session_mut(&mut self) -> &mut AsyncInstrumentSession {
        &mut self.session
    }

    fn dialect(&self) -> &'static instrument_core::DialectProfile {
        self.session.dialect_for(InstrumentKind::Switch)
    }

    fn cmd(&self, key: &str, vars: &[(&str, String)], fallback: String) -> String {
        dialect_io::try_formatted(self.dialect(), key, vars, fallback)
    }

    /// Formats a matrix path label for channels `ch1` and `ch2` (1-based).
    pub fn path_label(ch1: u32, ch2: u32) -> String {
        crate::classes::Switch::path_label(ch1, ch2)
    }

    /// Closes a route between two channels (1-based). IVI ClosePath equivalent.
    pub async fn close_route(&mut self, ch1: u32, ch2: u32) -> Result<()> {
        let cmd = self.cmd(
            "close_route",
            &[("ch1", ch1.to_string()), ("ch2", ch2.to_string())],
            scpi_commands::switch_close_route(ch1, ch2),
        );
        self.session.scpi_mut().write(&cmd).await
    }

    /// Opens a route between two channels (1-based).
    pub async fn open_route(&mut self, ch1: u32, ch2: u32) -> Result<()> {
        let cmd = self.cmd(
            "open_route",
            &[("ch1", ch1.to_string()), ("ch2", ch2.to_string())],
            scpi_commands::switch_open_route(ch1, ch2),
        );
        self.session.scpi_mut().write(&cmd).await
    }

    /// Returns whether a route is closed.
    pub async fn is_closed(&mut self, ch1: u32, ch2: u32) -> Result<bool> {
        let cmd = self.cmd(
            "is_closed",
            &[("ch1", ch1.to_string()), ("ch2", ch2.to_string())],
            scpi_commands::switch_is_closed(ch1, ch2),
        );
        let resp = self.session.scpi_mut().query(&cmd).await?;
        parse_closed(&resp)
    }

    /// Opens all routes.
    pub async fn open_all(&mut self) -> Result<()> {
        let cmd =
            dialect_io::try_command(self.dialect(), "open_all", scpi_commands::SWITCH_OPEN_ALL);
        self.session.scpi_mut().write(cmd).await
    }
}

File: dotnet/src/InstrumentComponents/Classes/AsyncSwitch.cs
using InstrumentComponents.Dialects;
using InstrumentComponents.Identity;
using InstrumentComponents.Kind;
using InstrumentComponents.Session;

namespace InstrumentComponents.Classes;

/// <summary>
/// Async switch / matrix session view.
/// Path model: routes are matrix channel pairs (ch1, ch2). Use <see cref="Switch.PathLabel"/> for naming;
/// IVI ClosePath maps to <see cref="CloseRouteAsync"/>.
/// </summary>
public sealed class AsyncSwitch
{
    private readonly AsyncInstrumentSession _session;

    public AsyncSwitch(AsyncInstrumentSession session) => _session = session;

    public AsyncInstrumentSession Session => _session;

    public Task<Idn> QueryIdnAsync(CancellationToken cancellationToken = default) =>
        _session.IdnAsync(cancellationToken);

    public Task ResetAsync(CancellationToken cancellationToken = default) =>
        _session.ResetAsync(cancellationToken);

    public Task OutputOffAsync(CancellationToken cancellationToken = default) =>
        OpenAllAsync(cancellationToken);

    private DialectProfile Dialect => _session.DialectFor(InstrumentKind.Switch);

    private string Cmd(string key, string fallback, params (string Name, string Value)[] vars) =>
        DialectCommand.Try(Dialect, key, fallback, vars);

    public static string PathLabel(uint ch1, uint ch2) => Switch.PathLabel(ch1, ch2);

    public Task CloseRouteAsync(uint ch1, uint ch2, CancellationToken cancellationToken = default) =>
        _session.Scpi.WriteAsync(Cmd("close_route", ScpiCommands.SwitchCloseRoute(ch1, ch2),
            ("ch1", ch1.ToString()), ("ch2", ch2.ToString())), cancellationToken);

    public Task OpenRouteAsync(uint ch1, uint ch2, CancellationToken cancellationToken = default) =>
        _session.Scpi.WriteAsync(Cmd("open_route", ScpiCommands.SwitchOpenRoute(ch1, ch2),
            ("ch1", ch1.ToString()), ("ch2", ch2.ToString())), cancellationToken);

    public async Task<bool> IsClosedAsync(uint ch1, uint ch2, CancellationToken cancellationToken = default)
    {
        var resp = await _session.Scpi.QueryAsync(Cmd("is_closed", ScpiCommands.SwitchIsClosed(ch1, ch2),
            ("ch1", ch1.ToString()), ("ch2", ch2.ToString())), cancellationToken).ConfigureAwait(false);
        return Switch.ParseClosed(resp);
    }

    public Task OpenAllAsync(CancellationToken cancellationToken = default) =>
        _session.Scpi.WriteAsync(Cmd("open_all", ScpiCommands.SwitchOpenAll), cancellationToken);
}
